### ISSUE TITLE
Linux support

### ISSUE THREAD

**Issue by tdhooper** [trevorvanhoof/sqrmelon#6]
Looking at the docs for this method, I can't see anything wrong with what you're doing. It's worth noting I had to make a few changes elsewhere to get to this point, see https://github.com/tdhooper/sqrmelon/commits/feature/ubuntu

```
Traceback (most recent call last):
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/sceneview3d.py", line 175, in paintGL
    _noSignalImage = loadImage(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'icons/nosignal.png'))
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/overlays/__init__.py", line 18, in loadImage
    glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_DECAL)
  File "/home/tdhooper/.local/lib/python2.7/site-packages/OpenGL/error.py", line 232, in glCheckError
    baseOperation = baseOperation,
OpenGL.error.GLError: GLError(
  err = 1282,
  description = 'invalid operation',
  baseOperation = glTexEnvf,
  cArguments = (
    GL_TEXTURE_ENV,
    GL_TEXTURE_ENV_MODE,
    GL_DECAL,
  )
)
```

**Comment by trevorvanhoof** [comment]
This call appeared unnecessary. I just removed it and that seems to work.
It does indeed seem to be used correctly, but possibly because it's old your previous GL version fix caused it to be non-existant in the core profile. I'd set the glFormat profile to QGLFormat.CompatibilityProfile! I also call glRecti which is deprecated in 4.5 core!

**Comment by tdhooper** [comment]
That's brought back the `GLSL 4.50 is not supported` error though. I don't think my machine supports 4.5 with the Compatibility Profile; on startup I get this:
```
X Error: GLXBadFBConfig 178
  Extension:    155 (Uknown extension)
  Minor opcode: 34 (Unknown request)
  Resource id:  0x4a00017
100.100
```

It's the same if I use `glFormat.setVersion(4, 5)`

Here's my glxinfo output:
```
sudo glxinfo | grep "OpenGL"
OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) HD Graphics 620 (Kaby Lake GT2) 
OpenGL core profile version string: 4.5 (Core Profile) Mesa 17.2.8
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.0 Mesa 17.2.8
OpenGL shading language version string: 1.30
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.2 Mesa 17.2.8
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
OpenGL ES profile extensions:
```

Thank you for looking into this, but seems like it's time for me to install windows :')

**Comment by trevorvanhoof** [comment]
Hey, I don't understand why it fails to set the version then. I don't think you actually need above 4.1 (though some fragment shaders define version 420 they don't depend on it as far as I'm aware).
Currently I try to initialize version 100.100 in hope that that makes GL select highest available, but I think I have to start another OpenGL context, ask highest available version, close it, and then start the actual program... Not sure when and how that's gonna happen yet, sorry!

**Comment by tdhooper** [comment]
I don't think that will fix it for me either, if I'm reading glxinfo right, my highest available version when not using the core profile is is 3.2

Anyway, yeah, I wouldn't worry about this any more, thank you for trying though :)

**Comment by trevorvanhoof** [comment]
I think that's just ES, GL Core should support 4.5 I think (looking at line 4)
But I'm guessing you get
"X Error: GLXBadFBConfig 178
  Extension:    155 (Uknown extension)
  Minor opcode: 34 (Unknown request)
  Resource id:  0x4a00017
4.5" <-- (this changed)
if you change the version in the code?

Well, I'll be happy to help if you ever decide to pick this up again. Maybe some day I can set up linux and work it all out myself :)

**Comment by tdhooper** [comment]
>But I'm guessing you get
"X Error: GLXBadFBConfig 178
Extension: 155 (Uknown extension)
Minor opcode: 34 (Unknown request)
Resource id: 0x4a00017
4.5" <-- (this changed)
if you change the version in the code?

Correct. I also get these:

````
RuntimeError: ('Shader compile failure (0): 0:1(10): error: GLSL 4.50 is not supported. Supported versions are: 1.10, 1.20, 1.30, 1.00 ES, 3.00 ES, 3.10 ES, and 3.20 ES\n', ['#version 450\nin vec2 vUV;uniform vec4 uColor;uniform sampler2D uImages[1];out vec4 outColor0;void main(){outColor0=uColor*texture(uImages[0], vUV);}'], GL_FRAGMENT_SHADER)
````

So I reckon 4.5 is only supported when I use the Core profile, not with the Compatibility profile. If I've got that right, the only options are to update the deprecated calls, or downgrade to 3.2.

**Comment by trevorvanhoof** [comment]
Hmm that is kinda weird, but at least I can reproduce :) I'll see if I can remove deprecated calls!

**Comment by trevorvanhoof** [comment]
Quick update: I still had a #version 450 for no reason in a default shader, now all shaders are #version 410, also upgraded legacy shaders and calls. Who knows now it'll magically work, probably now tho.

I also spent the day getting a virtual box with ubuntu set up, but it doesn't do accellerated graphics and just has software Mesa GL 3.3. 

After following various guides to get virtual box to use my GPU directly I am now somehow downgraded to chromium 2.1 instead and I give up on the virtual box idea!

**Comment by tdhooper** [comment]
That sounds painful, I hope your chromium is better now!

I tried the latest, and worked through a few things.

Firstly it was complaining about missing bloom.ixml, so I just removed it for now: https://github.com/tdhooper/sqrmelon/commit/1d4fe104e7d3e1103d15898864edb5f1da6ed4c1

Thanks to debugging code you added, I noticed that glFormat doesn't get used, we need to do this: https://github.com/tdhooper/sqrmelon/commit/2f2492625dbbab63f91976c645435d5c06495009

With that I change I got:
````
4.5 (Core Profile) Mesa 17.2.8
Traceback (most recent call last):
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/sceneview3d.py", line 186, in paintGL
    Scene.drawColorBufferToScreen(_noSignalImage, viewport)
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/scene.py", line 259, in drawColorBufferToScreen
    drawFullScreenRect()
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/scene.py", line 240, in drawFullScreenRect
    glDrawArrays(GL_TRIANGLE_FAN,0,4)
  File "/home/tdhooper/.local/lib/python2.7/site-packages/OpenGL/platform/baseplatform.py", line 402, in __call__
    return self( *args, **named )
  File "/home/tdhooper/.local/lib/python2.7/site-packages/OpenGL/error.py", line 232, in glCheckError
    baseOperation = baseOperation,
OpenGL.error.GLError: GLError(
	err = 1282,
	description = 'invalid operation',
	baseOperation = glDrawArrays,
	cArguments = (GL_TRIANGLE_FAN, 0, 4)
)
````

Which I managed to fix by binding a vertex array: https://github.com/tdhooper/sqrmelon/commit/2d9f76900323677d6ca99fcdeb5139a14a65380f

Now I see the 'No shot at current time' image, we're getting there! However when I add a shot and move the TimeSlider, I get the following error:

````
4.5 (Core Profile) Mesa 17.2.8
QFileSystemWatcher::addPaths: list is empty
QFileSystemWatcher::addPaths: list is empty
('Validation failure (0): ',)
Traceback (most recent call last):
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/timeslider.py", line 364, in mousePressEvent
    self._selectTime(event.x())
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/timeslider.py", line 335, in _selectTime
    self.__timer.time = x * (self.__timer.end - self.__timer.start) + self.__timer.start
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/timeslider.py", line 284, in time
    self.timeChanged.emit(value)
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/qtutil.py", line 131, in emit
    self.__connections[i](*args, **kwargs)
  File "__init__.py", line 400, in __setCurrentShot
    self.__sceneView.setScene(sc)
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/sceneview3d.py", line 111, in setScene
    self._scene.setSize(*self._size)
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/scene.py", line 519, in setSize
    self.colorBuffers[-1].append(Texture(Texture.RGBA32F, w, h, tile=value[3]))
  File "/home/tdhooper/Projects/sqrmelon/SqrMelon/buffers.py", line 98, in __init__
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP)
  File "/home/tdhooper/.local/lib/python2.7/site-packages/OpenGL/error.py", line 232, in glCheckError
    baseOperation = baseOperation,
OpenGL.error.GLError: GLError(
	err = 1280,
	description = 'invalid enumerant',
	baseOperation = glTexParameteri,
	cArguments = (
		GL_TEXTURE_2D,
		GL_TEXTURE_WRAP_S,
		GL_CLAMP,
	)
)
Warning: CGMath using pure-python fallback module.
````

**Comment by tdhooper** [comment]
Got it working :) #8

**Comment by trevorvanhoof** [comment]
I created a branch "linuxsupport" in which I pulled your changes.

I then moved some code around to avoid having to define fField everywhere (like you did in places like lensdirt.glsl).

Last I added back in the bloom.ixml (it is a comment in master.xml but it's a poor man's include system. the preprocessor will actually inline that file).

It runs on Windows again, but I'm wondering if I didn't break the actual linux support before merging back into master :) If you get a chance, please check it out & try to open a new project from this branch. Thanks for the work!

**Comment by trevorvanhoof** [comment]
Edit: I get invalid operations when drawing without binding anything (even though that should be valid afaik). Very interesting how an empty VAO is valid, I'd expect to have to allocate an empty GL_ARRAY_BUFFER or something like that. Binding a blank VAO would just reset all state.

So pushing a fix now to unbreak that and ignore this message!

-On top of the above I noticed there is a glBindVertexArray before drawFullScreenRect and a vertexArrayObject is being passed around, I wonder why that is necessary, as there is nothing stored inside the VAO as far as I know. I've removed it now (because if needed I hope to find a solution that doesn't involve forwarding the VAO through a deep call hierarchy, possibly by making it a static class-member)-

**Comment by tdhooper** [comment]
heh, yeah I don't understand it either, this is what led me to the solution though: https://stackoverflow.com/questions/8039929/opengl-drawarrays-without-binding-vbo

**Comment by tdhooper** [comment]
Looking good, needed a couple of fixes #9

**Comment by trevorvanhoof** [comment]
I wonder what the error was that caused copying bloom.ixml to Scenes
This should not be necessary and I think there then may be an issue with my xml include code
